File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        #initalize right and left product

        result = [1] * len(nums)

        left_product = 1
        for i in range(len(nums)):
            result[i] = left_product
            left_product *= nums[i]

        right_product = 1
        for i in range(len(nums) - 1, -1, -1):
            result[i] *= right_product
            right_product *= nums[i]

        return result
        




        left_product = 1





        
        

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) 
    {
        vector<int> res(nums.size(), 1);

        int left = 1;

        for (int i{}; i < nums.size(); i++)
        {
            res[i] = left;
            left *= nums[i];
        }

        printVector(res);

        int right = 1;

        for (int i = nums.size() - 1; i >= 0; i--)
        {
            res[i] *= right;
            right *= nums[i];
        }

        printVector(res);
        
        return res;
    }

    void printVector(const vector<int>& v) 
    {
        cout << "[";
        for (size_t i = 0; i < v.size(); ++i) {
            cout << v[i];
            if (i + 1 < v.size()) cout << ", ";
        }
        cout << "]" << endl;
    }

private:

};